File: src/game/spells/spells.ts
import type { EquipmentRow, GameDatabase, ItemRow } from '../data/types'
import { getEnchantment } from '../projects/projects'
import type { EquippedStack, PlayerSave } from '../save/types'

export const SPELL_SLOT_IDS = [
  'SLOT-0013',
  'SLOT-0014',
  'SLOT-0015',
  'SLOT-0016',
] as const

export type SpellSlotId = (typeof SPELL_SLOT_IDS)[number]

export function isSpellSlotId(slotId: string): slotId is SpellSlotId {
  return (SPELL_SLOT_IDS as readonly string[]).includes(slotId)
}

function capabilityTags(effects: string | null | undefined): string[] {
  if (typeof effects !== 'string') return []
  return effects
    .split(';')
    .map((part) => part.trim().toLowerCase())
    .filter(Boolean)
}

export function isSpellEquipment(equipment: EquipmentRow | undefined): boolean {
  if (!equipment) return false
  const tags = capabilityTags(equipment['Capabilities / Effects'])
  if (tags.includes('spell') || tags.some((tag) => tag.startsWith('spell_effect:'))) {
    return true
  }
  return isSpellSlotId(equipment['Slot ID'] ?? '')
}

export function isSpellItem(db: GameDatabase, itemId: string): boolean {
  const item = db.Items.find((row) => row['Item ID'] === itemId)
  if (item?.Category?.toLowerCase() === 'spell') return true
  const tags = String(item?.['Functional / Source Tags'] ?? '').toLowerCase()
  if (tags.includes('spell')) return true
  return isSpellEquipment(db.Equipment.find((row) => row['Item ID'] === itemId))
}

export function firstEmptySpellSlot(save: PlayerSave): SpellSlotId | null {
  for (const slotId of SPELL_SLOT_IDS) {
    if (!save.equipment.slots[slotId]?.itemId) return slotId
  }
  return null
}

/** All equipped spell stacks (empty slots omitted). Effects from each stack apply. */
export function equippedSpellStacks(save: PlayerSave): EquippedStack[] {
  const out: EquippedStack[] = []
  for (const slotId of SPELL_SLOT_IDS) {
    const stack = save.equipment.slots[slotId]
    if (stack?.itemId) out.push(stack)
  }
  return out
}

export function spellEffectEnchantmentId(
  db: GameDatabase,
  itemId: string,
): string | null {
  const equipment = db.Equipment.find((row) => row['Item ID'] === itemId)
  for (const tag of capabilityTags(equipment?.['Capabilities / Effects'])) {
    if (tag.startsWith('spell_effect:')) return tag.slice('spell_effect:'.length).toUpperCase()
  }
  const item = db.Items.find((row) => row['Item ID'] === itemId)
  const itemTags = String(item?.['Functional / Source Tags'] ?? '')
  for (const part of itemTags.split(';')) {
    const tag = part.trim().toLowerCase()
    if (tag.startsWith('spell_effect:')) {
      return tag.slice('spell_effect:'.length).toUpperCase()
    }
  }
  return null
}

function spellAllowsStacking(db: GameDatabase, itemId: string): boolean {
  const equipment = db.Equipment.find((row) => row['Item ID'] === itemId)
  const eqTags = capabilityTags(equipment?.['Capabilities / Effects'])
  if (eqTags.includes('spell_stacks')) return true
  if (eqTags.includes('spell_no_stack')) return false
  const item = db.Items.find((row) => row['Item ID'] === itemId)
  const itemTags = capabilityTags(item?.['Functional / Source Tags'])
  if (itemTags.includes('spell_stacks')) return true
  if (itemTags.includes('spell_no_stack')) return false
  // Default: stack (matches Strength / Abundance). Future uniques can opt out with spell_no_stack.
  return true
}

/**
 * Combine percent bonuses from equipped spells.
 * Stacking spells add; non-stacking spells contribute only their single best value.
 */
export function combineSpellPercentBonuses(
  contributions: Array<{ percent: number; stacks: boolean }>,
): number {
  let stacked = 0
  let bestUnique = 0
  for (const entry of contributions) {
    if (entry.percent <= 0) continue
    if (entry.stacks) stacked += entry.percent
    else bestUnique = Math.max(bestUnique, entry.percent)
  }
  return stacked + bestUnique
}

/** Damage-range bonus percent contributed by one spell item (0 if none). */
export function spellDamageRangeBonusPercent(db: GameDatabase, itemId: string): number {
  const equipment = db.Equipment.find((row) => row['Item ID'] === itemId)
  for (const tag of capabilityTags(equipment?.['Capabilities / Effects'])) {
    const match = tag.match(/^damage_range_bonus_percent:(\d+(?:\.\d+)?)$/)
    if (match) return Number(match[1])
  }

  const enchantmentId = spellEffectEnchantmentId(db, itemId)
  const enchantment = enchantmentId ? getEnchantment(db, enchantmentId) : undefined
  const effect = enchantment?.Effect ?? ''
  const match = effect.match(/\+(\d+(?:\.\d+)?)%\s*damage range/i)
  if (match) return Number(match[1])
  return 0
}

/** Chance percent to double item drop quantity from one Abundance-style spell. */
export function spellItemDoubleChancePercent(db: GameDatabase, itemId: string): number {
  const equipment = db.Equipment.find((row) => row['Item ID'] === itemId)
  for (const tag of capabilityTags(equipment?.['Capabilities / Effects'])) {
    const match = tag.match(/^item_double_chance_percent:(\d+(?:\.\d+)?)$/)
    if (match) return Number(match[1])
  }

  const enchantmentId = spellEffectEnchantmentId(db, itemId)
  const enchantment = enchantmentId ? getEnchantment(db, enchantmentId) : undefined
  const effect = enchantment?.Effect ?? ''
  const match = effect.match(
    /\+(\d+(?:\.\d+)?)%\s*chance to double item quantity/i,
  )
  if (match) return Number(match[1])
  return 0
}

/**
 * Multiplier from all equipped spells. Same bonus types add when spells stack
 * (2× Strength = +20% => 1.20). `nowMs` kept for call-site compatibility.
 */
export function activeSpellDamageRangeMultiplier(
  db: GameDatabase,
  save: PlayerSave,
  _nowMs: number = Date.now(),
): number {
  const contributions = equippedSpellStacks(save).map((stack) => ({
    percent: spellDamageRangeBonusPercent(db, stack.itemId),
    stacks: spellAllowsStacking(db, stack.itemId),
  }))
  return 1 + combineSpellPercentBonuses(contributions) / 100
}

/** Total chance to double item quantities on a successful drop (capped at 100). */
export function activeSpellItemDoubleChancePercent(
  db: GameDatabase,
  save: PlayerSave,
): number {
  const contributions = equippedSpellStacks(save).map((stack) => ({
    percent: spellItemDoubleChancePercent(db, stack.itemId),
    stacks: spellAllowsStacking(db, stack.itemId),
  }))
  return Math.min(100, combineSpellPercentBonuses(contributions))
}

export function spellTooltipLines(
  db: GameDatabase,
  item: ItemRow | undefined,
  itemId: string,
): string[] {
  const lines: string[] = []
  const enchantmentId = spellEffectEnchantmentId(db, itemId)
  const enchantment = enchantmentId ? getEnchantment(db, enchantmentId) : undefined
  if (enchantment?.['Display Name']) lines.push(enchantment['Display Name'])
  if (enchantment?.Effect) lines.push(enchantment.Effect)
  else if (item?.Description) lines.push(item.Description)
  lines.push('Always active while equipped.')
  if (spellAllowsStacking(db, itemId)) {
    lines.push('Duplicate copies stack.')
  } else {
    lines.push('Does not stack with duplicate copies.')
  }
  return lines
}

File: packages/ik_rules/lib/src/spells/spells.dart
import 'dart:math' as math;

import 'package:collection/collection.dart';
import 'package:ik_content/ik_content.dart';

import '../js_compat.dart';
import '../projects/projects.dart';
import '../save/generated/save_models.dart';
import '../tags.dart';

const List<String> spellSlotIds = <String>['SLOT-0013', 'SLOT-0014', 'SLOT-0015', 'SLOT-0016'];

bool isSpellSlotId(String slotId) => spellSlotIds.contains(slotId);

bool isSpellEquipment(EquipmentRow? equipment) {
  if (equipment == null) return false;
  final tags = capabilityTags(equipment.raw['Capabilities / Effects']);
  if (tags.contains('spell') || tags.any((tag) => tag.startsWith('spell_effect:'))) {
    return true;
  }
  final slotId = equipment.raw['Slot ID'];
  return isSpellSlotId(slotId is String ? slotId : '');
}

bool isSpellItem(GameDatabase db, String itemId) {
  final item = db.items.firstWhereOrNull((row) => row.raw['Item ID'] == itemId);
  if (lowerOrEmpty(item?.raw['Category']) == 'spell') return true;
  final tags = jsString(item?.raw['Functional / Source Tags'] ?? '').toLowerCase();
  if (tags.contains('spell')) return true;
  return isSpellEquipment(db.equipment.firstWhereOrNull((row) => row.raw['Item ID'] == itemId));
}

String? firstEmptySpellSlot(PlayerSave save) {
  for (final slotId in spellSlotIds) {
    if (isBlank(save.equipment.slots[slotId]?.itemId)) return slotId;
  }
  return null;
}

/// All equipped spell stacks (empty slots omitted). Effects from each apply.
List<EquippedStack> equippedSpellStacks(PlayerSave save) {
  final out = <EquippedStack>[];
  for (final slotId in spellSlotIds) {
    final stack = save.equipment.slots[slotId];
    if (stack != null && isNotBlank(stack.itemId)) out.add(stack);
  }
  return out;
}

String? spellEffectEnchantmentId(GameDatabase db, String itemId) {
  final equipment = db.equipment.firstWhereOrNull((row) => row.raw['Item ID'] == itemId);
  for (final tag in capabilityTags(equipment?.raw['Capabilities / Effects'])) {
    if (tag.startsWith('spell_effect:')) {
      return tag.substring('spell_effect:'.length).toUpperCase();
    }
  }
  final itemTags = jsString(
    db.items
            .firstWhereOrNull((row) => row.raw['Item ID'] == itemId)
            ?.raw['Functional / Source Tags'] ??
        '',
  );
  for (final part in itemTags.split(';')) {
    final tag = part.trim().toLowerCase();
    if (tag.startsWith('spell_effect:')) {
      return tag.substring('spell_effect:'.length).toUpperCase();
    }
  }
  return null;
}

bool _spellAllowsStacking(GameDatabase db, String itemId) {
  final equipment = db.equipment.firstWhereOrNull((row) => row.raw['Item ID'] == itemId);
  final equipmentTags = capabilityTags(equipment?.raw['Capabilities / Effects']);
  if (equipmentTags.contains('spell_stacks')) return true;
  if (equipmentTags.contains('spell_no_stack')) return false;
  final item = db.items.firstWhereOrNull((row) => row.raw['Item ID'] == itemId);
  final itemTags = capabilityTags(item?.raw['Functional / Source Tags']);
  if (itemTags.contains('spell_stacks')) return true;
  if (itemTags.contains('spell_no_stack')) return false;
  // Default: stack (matches Strength / Abundance). Future uniques opt out with spell_no_stack.
  return true;
}

/// One spell's contribution to a percent bonus.
class SpellContribution {
  const SpellContribution({required this.percent, required this.stacks});

  final num percent;
  final bool stacks;
}

/// Combines percent bonuses from equipped spells.
///
/// Stacking spells add; non-stacking spells contribute only their best value.
num combineSpellPercentBonuses(List<SpellContribution> contributions) {
  num stacked = 0;
  num bestUnique = 0;
  for (final entry in contributions) {
    if (entry.percent <= 0) continue;
    if (entry.stacks) {
      stacked += entry.percent;
    } else {
      bestUnique = math.max(bestUnique, entry.percent);
    }
  }
  return stacked + bestUnique;
}

String _enchantmentEffect(GameDatabase db, String itemId) {
  final enchantmentId = spellEffectEnchantmentId(db, itemId);
  if (isBlank(enchantmentId)) return '';
  final effect = getEnchantment(db, enchantmentId!)?.raw['Effect'];
  return effect is String ? effect : '';
}

/// Damage-range bonus percent contributed by one spell item (0 if none).
num spellDamageRangeBonusPercent(GameDatabase db, String itemId) {
  final equipment = db.equipment.firstWhereOrNull((row) => row.raw['Item ID'] == itemId);
  for (final tag in capabilityTags(equipment?.raw['Capabilities / Effects'])) {
    final match = RegExp(r'^damage_range_bonus_percent:(\d+(?:\.\d+)?)$').firstMatch(tag);
    if (match != null) return jsNumber(match.group(1));
  }

  final match = RegExp(
    r'\+(\d+(?:\.\d+)?)%\s*damage range',
    caseSensitive: false,
  ).firstMatch(_enchantmentEffect(db, itemId));
  return match == null ? 0 : jsNumber(match.group(1));
}

/// Chance percent to double item drop quantity from one Abundance-style spell.
num spellItemDoubleChancePercent(GameDatabase db, String itemId) {
  final equipment = db.equipment.firstWhereOrNull((row) => row.raw['Item ID'] == itemId);
  for (final tag in capabilityTags(equipment?.raw['Capabilities / Effects'])) {
    final match = RegExp(r'^item_double_chance_percent:(\d+(?:\.\d+)?)$').firstMatch(tag);
    if (match != null) return jsNumber(match.group(1));
  }

  final match = RegExp(
    r'\+(\d+(?:\.\d+)?)%\s*chance to double item quantity',
    caseSensitive: false,
  ).firstMatch(_enchantmentEffect(db, itemId));
  return match == null ? 0 : jsNumber(match.group(1));
}

/// Multiplier from all equipped spells. Same bonus types add when spells stack
/// (2× Strength = +20% => 1.20).
num activeSpellDamageRangeMultiplier(GameDatabase db, PlayerSave save) {
  final contributions = equippedSpellStacks(save)
      .map(
        (stack) => SpellContribution(
          percent: spellDamageRangeBonusPercent(db, stack.itemId),
          stacks: _spellAllowsStacking(db, stack.itemId),
        ),
      )
      .toList();
  return 1 + combineSpellPercentBonuses(contributions) / 100;
}

/// Total chance to double item quantities on a successful drop (capped at 100).
num activeSpellItemDoubleChancePercent(GameDatabase db, PlayerSave save) {
  final contributions = equippedSpellStacks(save)
      .map(
        (stack) => SpellContribution(
          percent: spellItemDoubleChancePercent(db, stack.itemId),
          stacks: _spellAllowsStacking(db, stack.itemId),
        ),
      )
      .toList();
  return math.min(100, combineSpellPercentBonuses(contributions));
}

List<String> spellTooltipLines(GameDatabase db, ItemRow? item, String itemId) {
  final lines = <String>[];
  final enchantmentId = spellEffectEnchantmentId(db, itemId);
  final enchantment = isBlank(enchantmentId) ? null : getEnchantment(db, enchantmentId!);
  final displayName = enchantment?.raw['Display Name'];
  final effect = enchantment?.raw['Effect'];
  final description = item?.raw['Description'];
  if (displayName is String && displayName.isNotEmpty) lines.add(displayName);
  if (effect is String && effect.isNotEmpty) {
    lines.add(effect);
  } else if (description is String && description.isNotEmpty) {
    lines.add(description);
  }
  lines.add('Always active while equipped.');
  lines.add(
    _spellAllowsStacking(db, itemId)
        ? 'Duplicate copies stack.'
        : 'Does not stack with duplicate copies.',
  );
  return lines;
}
